### PR TITLE
Shader optimizations to reduce VGPR usage and increase occupancy

### PR DESCRIPTION
**Pull request by reduz**
Reduced from 116 to 80 VGPRs in normal usage, still more work is needed.

Some not often used features are gone for the sake of this and some are disabled, will re-enable later.
-Shadow color per light is gone
-Oren Nayar is gone (implement manually if you really want it, makes little difference)
-SSAO color is gone
-SSAO ao affect is gone because this confused users, min(ao,ssao) now used, should work in most cases.
Also

-softshadows (based on angular distance) are likely going to be removed from directional light, too expensive dont look great, and dont make much of a difference
-projectors are disabled.

### CODE REVIEW

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp [hunk 1]
```diff
@@ -6288,13 +6283,14 @@ void RendererSceneRenderRD::_setup_lights(const PagedArray<RID> &p_lights, const
 				float sign = storage->light_is_negative(base) ? -1 : 1;
 				Color linear_col = storage->light_get_color(base).to_linear();
 
-				light_data.attenuation_energy[0] = Math::make_half_float(storage->light_get_param(base, RS::LIGHT_PARAM_ATTENUATION));
-				light_data.attenuation_energy[1] = Math::make_half_float(sign * storage->light_get_param(base, RS::LIGHT_PARAM_ENERGY) * Math_PI);
+				light_data.attenuation = storage->light_get_param(base, RS::LIGHT_PARAM_ATTENUATION);
```
Comment (reduz) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:6286]: This just cleaned up how parameters are sent to the reflection probe struct

File: servers/rendering/renderer_rd/renderer_scene_render_rd.h [hunk 2]
```diff
@@ -1313,10 +1322,15 @@ class RendererSceneRenderRD : public RendererSceneRender {
 			float inv_radius;
 			float direction[3];
 			float size;
-			uint16_t attenuation_energy[2]; //16 bits attenuation, then energy
-			uint8_t color_specular[4]; //rgb color, a specular (8 bit unorm)
-			uint16_t cone_attenuation_angle[2]; // attenuation and angle, (16bit float)
-			uint8_t shadow_color_enabled[4]; //shadow rgb color, a>0.5 enabled (8bit unorm)
+
```
Comment (reduz) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.h:1325]: Likewise cleaned up lights

File: servers/rendering/renderer_rd/shaders/scene_forward.glsl [hunk 3]
```diff
@@ -591,14 +598,25 @@ LIGHT_SHADER_CODE
...
+#else
+	float cLdotH = clamp(dot(L, H), 0.0, 1.0);
+#endif
 #endif
 
+	float metallic = unpackUnorm4x8(orms).z;
 	if (metallic < 1.0) {
+		float roughness = unpackUnorm4x8(orms).y;
+
 #if defined(DIFFUSE_OREN_NAYAR)
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/scene_forward.glsl:620]: If deleting Oren-Nayar, it should be deleted here and in the material and shader as well

File: servers/rendering/renderer_rd/shaders/scene_forward.glsl [hunk 4]
```diff
@@ -2710,173 +3009,149 @@ FRAGMENT_SHADER_CODE
...
-	if (scene_data.ssao_enabled) {
-		float ao = texture(sampler2D(ao_buffer, material_samplers[SAMPLER_LINEAR_CLAMP]), screen_uv).r;
-		ambient_light = mix(scene_data.ao_color.rgb, ambient_light, ao);
-		float ao_light_affect = mix(1.0, ao, scene_data.ssao_light_affect);
-		specular_light = mix(scene_data.ao_color.rgb, specular_light, ao_light_affect);
-		diffuse_light = mix(scene_data.ao_color.rgb, diffuse_light, ao_light_affect);
-	}
-#endif //LOW_END_MODE
+	// multiply by albedo
+	diffuse_light *= albedo; // ambient must be multiplied by albedo at the end
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/scene_forward.glsl:3140]: comment out of date

File: servers/rendering/renderer_rd/shaders/scene_forward.glsl [hunk 5]
```diff
@@ -2297,6 +2335,14 @@ FRAGMENT_SHADER_CODE
 	}
 #endif
 
+#ifndef LOW_END_MODE
+	if (scene_data.ssao_enabled) {
+		float ssao = texture(sampler2D(ao_buffer, material_samplers[SAMPLER_LINEAR_CLAMP]), screen_uv).r;
+		ao = min(ao, ssao);
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/scene_forward.glsl:2341]: We need to delete SSAO_ao_affect from environment etc.

**Review by clayjohn — COMMENTED** [review]
Looks mostly good!

**Comment by clayjohn** [comment]
Im getting 2 types of artifacts with shaders. One is random blocks of shadow that flicker in and out (likely entire clusters missing a light). And the second is that the derivative shadow optimization doesnt scale well with distance from a shadow. So when you get far enough away from a shadow the edges take on a very aliased look

Screenshot (60)

File: servers/rendering/renderer_rd/shaders/cluster_render.glsl [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+		//http://advances.realtimerendering.com/s2017/2017_Sig_Improved_Culling_final.pdf
+
+		uvec4 mask;
+
+		while (true) {
+			// find the cluster offset of the first active thread
+			// threads that did break; go inactive and no longer count
+			uint first = subgroupBroadcastFirst(cluster_offset);
+			// update the mask for thread that match this cluster
+			uvec4 mask = subgroupBallot(first == cluster_offset);
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/cluster_render.glsl:129]:
```suggestion
			mask = subgroupBallot(first == cluster_offset);
```

This line is causing the issue that I posted about.

Comment (reduz) [reply servers/rendering/renderer_rd/shaders/cluster_render.glsl]: you are a genius, thanks! fixed.

**Comment by reduz** [comment]
@akien-mga Should hopefully be ok to merge but the static checks are not passing and I already have clang-format-11.

**Comment by Calinou** [comment]
@reduz I would recommend avoiding nested ternary operators, it should fix the clang-format issue at the same time.

**Comment by clayjohn** [comment]
I haven't reviewed the code yet, but running this build there are some artifacts created on triangle edges. It must be due to the shadow derivative path.

Screenshot (72)

**Comment by reduz** [comment]
@clayjohn disabled it for now

**Comment by JFonS** [comment]
I went over the files, it wasn't an in-depth review but I didn't spot anything wrong.

Other than that, Github lists some files as edited, but they don't contain any changes, not sure what's going on. Also, static checks have failed, I think they should be restarted.

**Comment by akien-mga** [comment]
I restarted CI jobs for static checks and Android which failed for a seemingly unrelated reason, likely a CI oops.

**Comment by akien-mga** [comment]
> I restarted CI jobs for static checks and Android which failed for a seemingly unrelated reason, likely a CI oops.

Ah no it's not a CI issue, @reduz did remove executable permission on those files, so they can't run on CI:
Screenshot_20210119_231914

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: servers/rendering/renderer_rd/shaders/scene_forward.glsl (hunk 3 above)
Comment (Chaosus) [reply servers/rendering/renderer_rd/shaders/scene_forward.glsl:620]: Fixed by https://github.com/godotengine/godot/pull/49419

File: servers/rendering/renderer_rd/shaders/scene_forward.glsl (hunk 5 above)
Comment (Calinou) [reply servers/rendering/renderer_rd/shaders/scene_forward.glsl:2408]: For reference, I opened an issue to track this: https://github.com/godotengine/godot/issues/54611

**Pull request merged**